test/libsolidity/util/TestFileParser.cpp

```diff
@@ -0,0 +1,385 @@
+
+#include <test/Options.h>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/optional.hpp>
+#include <boost/throw_exception.hpp>
+#include <fstream>
+#include <memory>
+#include <stdexcept>
+
+using namespace dev;
+using namespace langutil;
+using namespace solidity;
+using namespace dev::solidity::test;
+using namespace std;
+
+namespace
+{
+	bool isIdentifierStart(char c)
+	{
+		return c == '_' || c == '$' || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z');
+	}
+	bool isIdentifierPart(char c)
+	{
+		return isIdentifierStart(c) || isdigit(c);
+	}
+}
+
+vector<dev::solidity::test::FunctionCall> TestFileParser::parseFunctionCalls()
+{
+	vector<FunctionCall> calls;
+	if (!accept(SoltToken::EOS))
+	{
+		assert(m_scanner.currentToken() == SoltToken::Unknown);
+		m_scanner.scanNextToken();
+
+		while (!accept(SoltToken::EOS))
+		{
+			if (!accept(SoltToken::Whitespace))
+			{
+				FunctionCall call;
+
+				/// If this is not the first call in the test,
+				/// the last call to parseParameter could have eaten the
+				/// new line already. This could only be fixed with a one
+				/// token lookahead that checks parseParameter
+				/// if the next token is an identifier.
+				if (calls.empty())
+					expect(SoltToken::Newline);
+				else
+					accept(SoltToken::Newline, true);
+
+				call.signature = parseFunctionSignature();
+				if (accept(SoltToken::Comma, true))
+					call.value = parseFunctionCallValue();
+				if (accept(SoltToken::Colon, true))
+					call.arguments = parseFunctionCallArguments();
+
+				if (accept(SoltToken::Newline, true))
+					call.displayMode = FunctionCall::DisplayMode::MultiLine;
+
+				call.arguments.comment = parseComment();
+
+				if (accept(SoltToken::Newline, true))
+					call.displayMode = FunctionCall::DisplayMode::MultiLine;
+
+				expect(SoltToken::Arrow);
+				call.expectations = parseFunctionCallExpectations();
+				call.expectations.comment = parseComment();
+
+				calls.emplace_back(std::move(call));
+			}
+			else
+				m_scanner.scanNextToken();
+		}
+	}
+	return calls;
+}
+
+string TestFileParser::formatToken(SoltToken _token)
+{
+	switch (_token)
+	{
+#define T(name, string, precedence) case SoltToken::name: return string;
+		SOLT_TOKEN_LIST(T, T)
+#undef T
+		default: // Token::NUM_TOKENS:
+			return "";
+	}
+}
+
+bool TestFileParser::accept(SoltToken _token, bool const _expect)
+{
+	if (m_scanner.currentToken() == _token)
+	{
+		if (_expect)
+			expect(_token);
+		return true;
+	}
+	return false;
+}
+
+bool TestFileParser::expect(SoltToken _token, bool const _advance)
+{
+	if (m_scanner.currentToken() != _token)
+		throw Error(
+			Error::Type::ParserError,
+			"Unexpected " + formatToken(m_scanner.currentToken()) + ": \"" +
+			m_scanner.currentLiteral() + "\". " +
+			"Expected \"" + formatToken(_token) + "\"."
+			);
+	if (_advance)
+		m_scanner.scanNextToken();
+	return true;
+}
+
+string TestFileParser::parseFunctionSignature()
```

Comment:
Well, this now won't work for signatures involving structs... In the ABI structs are considered tuples, so for ``f(S memory s, uint256 b) public pure {}`` with ``struct S { uint256 a; bytes32 b; }`` the signature would be ``f((uint256,bytes32),uint256)``.

I would be fine with skipping this for now, but we should at least add a TODO comment about that here. It would be straightforward to count opening and closing parentheses, though, so we could just do it now, as you like... It's not that important, because I think we don't have any test case at the moment that would use it - on the other hand it would be really handy for #1603 of course :-D.

Reply:
Well, function signature do now allow recursive tuples e.g. `f(((((bytes, bytes, (bytes)), bytes), bytes), (bytes, bytes)), (bytes, bytes))`. I thought that might that might be useful.
Let's get this merged soon, https://github.com/ethereum/solidity/pull/5872 after and thenyou'll be able to run those tests ;)